pulp_rpm/app/tasks/comps.py

```diff
@@ -26,9 +29,15 @@ def parse_comps_components(comps_file):
     comps = libcomps.Comps()
     # Read the file and pass the string along because comps.fromxml_f() will only take a
     # path-string that doesn't work on things like S3 storage
-    with comps_file.file.open("rb") as f:
-        data = f.read()
-        comps.fromxml_str(data.decode("utf-8"))
+    with comps_file.file.open("rb") as comps_uploaded:
```

Comment:
So the sequence here is : 

* Incoming is comps_file which is a PulpTemporaryFile
* We copy it to a NamedTemporaryFile in "." (ie the task's working-directory)
* We uncompress it (possibly, since it may not be compressed to start with) into "groups.xml" (shouldn't that be comps.xml? Does it matter?)
* Then we open that (possibly) uncompressed file, and read it into a string, so we can pass **that** to fromxml_str

I'm unclear on why we need to go from comps_file to comps_on_disk. And if we do - can we use os.sendfile() or shutil.copyfile()  (or one of its friends) here instead of read/write/flush? Is there a reason not to?

Reply:
>We uncompress it (possibly, since it may not be compressed to start with) into "groups.xml" (shouldn't that be comps.xml? Does it matter?)

I don't think it matters but you're right anyway, that would make sense.

>I'm unclear on why we need to go from comps_file to comps_on_disk. And if we do - can we use os.sendfile() or shutil.copyfile() (or one of its friends) here instead of read/write/flush? Is there a reason not to?

PulpTemporaryFile as it exists currently is backed by the Pulp artifact store, which could be in S3.